### PR TITLE
[GH-898] Separate mandatory and optional inputs.

### PR DESCRIPTION
**Pull request by rexwangcc**
### Purpose

- https://broadinstitute.atlassian.net/browse/GH-898

### Changes

- Remove the confusing default chip-metadata inputs from the codebase.
- Genotype concordance inputs are optional now.
- Add unit tests accordingly.

### Review Instructions

- No instructions.

### CODE REVIEW

File: api/test/zero/unit/aou_module_test.clj [hunk 1]
```diff
@@ -23,43 +23,50 @@
...
+                                                  :Arrays.call_rate_threshold
+                                                  :Arrays.haplotype_database_file])
+        all-expected-keys                   (sort (concat all-expected-keys-no-control
+                                                          [:Arrays.control_sample_vcf_index_file
+                                                           :Arrays.control_sample_intervals_file
+                                                           :Arrays.control_sample_vcf_file
+                                                           :Arrays.control_sample_name]))]
     (testing "aou filters out non-necessary keys for per-sample-inputs"
-      (is (= (aou/get-per-sample-inputs redundant-per-sample-inputs-inputs) expected-per-sample-inputs)))
+      (is (= expected-per-sample-inputs (aou/get-per-sample-inputs redundant-per-sample-inputs-inputs))))
```
Comment (rexwangcc) [inline api/test/zero/unit/aou_module_test.clj:62]: > Don't talk like yoda. This kaocha plugin lets you write (is (= actual expected)).

Unfortunately, we are the Mandalorians.

File: api/src/zero/module/aou.clj [hunk 2]
```diff
@@ -47,13 +38,6 @@
...
-(def genotype-concordance
-  "Genotype Concordance inputs for arrays."
-  {:control_sample_vcf_file       "gs://broad-gotc-test-storage/arrays/controldata/NA12878.vcf.gz"
-   :control_sample_vcf_index_file "gs://broad-gotc-test-storage/arrays/controldata/NA12878.vcf.gz.tbi"
-   :control_sample_intervals_file "gs://broad-gotc-test-storage/arrays/controldata/NA12878.interval_list"
-   :control_sample_name           "NA12878"})
-
 (def other-inputs
   "Miscellaneous inputs for arrays."
   {:call_rate_threshold              0.98
```
Comment (rexwangcc) [inline api/src/zero/module/aou.clj:43]: Still not sure about `:genotype_concordance_threshold`, @gbggrant could you help check if we could always pass these by default?

File: api/test/zero/unit/aou_module_test.clj (hunk 1 above)
Comment (samanehsan) [reply api/test/zero/unit/aou_module_test.clj:62]: 🤣

**Comment by samanehsan** [comment]
Also I made a ticket to track this change: https://broadinstitute.atlassian.net/jira/software/projects/GH/boards/530?selectedIssue=GH-898

File: api/src/zero/module/aou.clj [hunk 3]
```diff
@@ -30,15 +30,6 @@
   (let [[key value] (first cromwell-label-map)]
     (str (name key) ":" value)))
 
-(def chip-metadata
```
Comment (jack-r-warren) [inline api/src/zero/module/aou.clj:33]: I'm not up to date on the discussion on this. So long as we really don't need this anymore even as a default this all looks good.

File: api/src/zero/module/aou.clj (hunk 2 above)
Comment (gbggrant) [reply api/src/zero/module/aou.clj:43]: The genotype_concordance_threshold can be removed - the WDL defaults to 0.98 and it isn't normally set externally.
However the call_rate_threshold is set in the LIMS message and correspondingly in the arrays message.

**Comment by rexwangcc** [comment]
Thanks @gbggrant ! 

Does it make sense to set `call_rate_threshold` to 0.98 by default and let JMS message's values to overwrite it per sample, or it's a mandatory field and should NOT have default values?

@samanehsan I don't see `call_rate_threshold ` field in the payload of the request to WFL, but I do see `callRateThreshold` in the JMS message... We probably need to look into where it gets dropped...

File: api/test/zero/unit/aou_module_test.clj [hunk 4]
```diff
@@ -23,43 +24,49 @@
...
-                                            :Arrays.control_sample_vcf_index_file
-                                            :Arrays.genotype_concordance_threshold
-                                            :Arrays.ref_fasta
-                                            :Arrays.bead_pool_manifest_file
-                                            :Arrays.analysis_version_number
-                                            :Arrays.fingerprint_genotypes_vcf_file
-                                            :Arrays.control_sample_name
-                                            :Arrays.cluster_file
-                                            :Arrays.call_rate_threshold
-                                            :Arrays.haplotype_database_file]]
```
Comment (tbl3rd) [inline api/test/zero/unit/aou_module_test.clj:59]: Why `sort`?

File: api/test/zero/unit/aou_module_test.clj [hunk 5]
```diff
@@ -23,43 +24,49 @@
                                             :zcall_thresholds_file       "foo"}
```
Comment (tbl3rd) [inline api/test/zero/unit/aou_module_test.clj:24]: That's a lot of `"foo"`.

File: api/test/zero/unit/aou_module_test.clj [hunk 6]
```diff
@@ -23,43 +24,49 @@
...
       (is (thrown? Exception (aou/get-per-sample-inputs missing-per-sample-inputs-inputs))))
     (testing "aou prepares all necessary keys"
-      (is (= (keys (aou/make-inputs :aou-dev expected-per-sample-inputs)) all-expected-keys)))))
+      (is (= all-expected-keys-no-control (sort (keys (aou/make-inputs :aou-dev expected-per-sample-inputs))))))
+    (testing "aou prepares all necessary keys plus optional keys"
+      (is (= all-expected-keys (sort (keys (aou/make-inputs :aou-dev (merge expected-per-sample-inputs
+                                                                       {:control_sample_vcf_index_file "foo"
+                                                                        :control_sample_intervals_file "foo"
+                                                                        :control_sample_vcf_file       "foo"
+                                                                        :control_sample_name           "foo"})))))))))
```
Comment (tbl3rd) [inline api/test/zero/unit/aou_module_test.clj:72]: Ah.  Maybe use `set`s instead?

File: api/test/zero/unit/aou_module_test.clj [hunk 7]
```diff
@@ -23,43 +24,49 @@
...
+                                                  :Arrays.analysis_version_number
+                                                  :Arrays.fingerprint_genotypes_vcf_file
+                                                  :Arrays.cluster_file
+                                                  :Arrays.call_rate_threshold
+                                                  :Arrays.haplotype_database_file])
+        all-expected-keys                   (sort (concat all-expected-keys-no-control
+                                                          [:Arrays.control_sample_vcf_index_file
+                                                           :Arrays.control_sample_intervals_file
+                                                           :Arrays.control_sample_vcf_file
+                                                           :Arrays.control_sample_name]))]
```
Comment (tbl3rd) [inline api/test/zero/unit/aou_module_test.clj:60]: Seems like we could compute this, but maybe we don't wanna?

**Review by ehigham — APPROVED** [review]

File: api/test/zero/unit/aou_module_test.clj (hunk 7 above)
Comment (rexwangcc) [reply api/test/zero/unit/aou_module_test.clj:60]: yeah, we could derive this, i was just being lazy and in a hurry so didn't do so.

File: api/test/zero/unit/aou_module_test.clj (hunk 4 above)
Comment (rexwangcc) [reply api/test/zero/unit/aou_module_test.clj:59]: since it seems the order matters with `[]`

**Comment by rexwangcc** [comment]
> I assume there is another PR to augment the `cromwell-label-map`?

That PR was https://github.com/broadinstitute/wfl/pull/106/files, it didn't parameterize `cromwell-label-map` but instead add a `make-labels` function. That was merged before I saw your comment that we could use the keys extracted from inputs...

File: api/test/zero/unit/aou_module_test.clj (hunk 4 above)
Comment (tbl3rd) [reply api/test/zero/unit/aou_module_test.clj:59]: Order matters for `seq`s but not for `set`s.  Hence my comment below.

File: api/test/zero/unit/aou_module_test.clj (hunk 7 above)
Comment (tbl3rd) [reply api/test/zero/unit/aou_module_test.clj:60]: OK by me.

File: api/test/zero/unit/aou_module_test.clj (hunk 4 above)
Comment (rexwangcc) [reply api/test/zero/unit/aou_module_test.clj:59]: Fair enough, changing them to `set` now.

**Pull request merged**